traitlets/traitlets.py

```diff
@@ -489,11 +488,28 @@ def _validate(self, obj, value):
         return value
 
     def _cross_validate(self, obj, value):
-        if hasattr(obj, '_%s_validate' % self.name):
-            cross_validate = getattr(obj, '_%s_validate' % self.name)
-            value = cross_validate(value, self)
+        try:
+            cb = getattr(obj, '_%s_validate' % self.name)
+        except AttributeError:
+            cb = obj._trait_validators.get(self.name,None)
+        else:
+            warn("_[traitname]_validate handlers are deprecated: use register_validator instead",
+                 DeprecationWarning, stacklevel=2)
+            if not callable(cb):
```

Comment:
I am fine with plain functions being registered as validators using a
`register_validator` method.

On Wed, Sep 2, 2015 at 2:56 PM, Jason Grout <notifications@github.com>
wrote:

> In traitlets/traitlets.py
> <https://github.com/ipython/traitlets/pull/73#discussion_r38590023>:
>
> > @@ -489,11 +488,28 @@ def _validate(self, obj, value):
> >          return value
> >
> >      def _cross_validate(self, obj, value):
> > -        if hasattr(obj, '_%s_validate' % self.name):
> > -            cross_validate = getattr(obj, '_%s_validate' % self.name)
> > -            value = cross_validate(value, self)
> > +        try:
> > +            cb = getattr(obj, '_%s_validate' % self.name)
> > +        except AttributeError:
> > +            cb = obj._trait_validators.get(self.name,None)
> > +        else:
> > +            warn("_[traitname]_validate handlers are deprecated: use register_validator instead",
> > +                 DeprecationWarning, stacklevel=2)
> > +            if not callable(cb):
>
> Thanks for pointing out the consistency issues. Personally, I would
> eliminate the extra code checking for callable. It makes the code simpler,
> and essentially the same error will be raised by python automatically if it
> isn't callable.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ipython/traitlets/pull/73/files#r38590023>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com